### PR TITLE
Solid properties module

### PR DESCRIPTION
**Pull request by aprilnovak**
This PR includes a draft for a solid properties module. The design is essentially identical to the fluid property module, i.e. UserObjects are used to define property implementations that are then used for computing material properties in Materials. Four materials are implemented:

- Stainless steel 316
- Graphite
- Silicon carbide
- Function-based properties for arbitrary custom specification

This formulation permits solid thermal properties to be declared as a function of temperature only. Some applications may require solid properties as a function of other variables, such as porosity and burnup. To accommodate the dependence on an arbitrary number of dependent variables would require a proliferation of base classes each with a different interface function, i.e. `k_from_T(Real T)` and `k_from_T_burnup(Real T, Real burnup)`, ... .

This PR is marked as WIP because a `Material`-based route would permit every solid to require whatever coupled variables are needed for property evaluation. But, materials are not computed at nodes, and applications requiring materials in nodal AuxKernels would not work with the `Material`-based route (from my understanding at least). 

I'm inclined to go with the `Material` route because it seems more flexible for accommodating more applications, but I also want to propose the `UserObject` route because I am unfamiliar with the general consensus on how pervasive the use of material properties in nodal `AuxKernels` in other applications might be. 

Thoughts? I'll change the design to the `Material`-based design if others agree.

Refs #12284

### CODE REVIEW

**Comment by aprilnovak** [comment]
Also, I am still trying to figure out why some of the documentation won't build (syntax can't be found?), so building the documentation isn't the prettiest yet

**Comment by permcody** [comment]
It probably means that one of your links isn't valid. Remember that links are found in the source folder and the capitalization might be different. 

Remove the [WIP] so that you trigger a build then we can see the documentation target and may be able to help you with specifics.

File: modules/solid_properties/src/userobjects/ThermalGraphiteProperties.C [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+  if (T > 373.15)
+    term = (2.1e-9 * std::pow(T, 3.0) - 1.23726e-5 * std::pow(T, 2.0) +
+      3.05359e-2 * T - 9.73349) * 1e-7;
+
+  return _beta0 + term;
+}
+
+Real
+ThermalGraphiteProperties::emissivity_from_T(Real T) const
```
Comment (YaqiWang) [inline modules/solid_properties/src/userobjects/ThermalGraphiteProperties.C]: This function looks to be *surface* emissivity different from the volumetric emissivity appeared in thermal radiation transport equation. Is this function used for imposing boundary condition of the thermal conduction equation?

Comment (YaqiWang) [reply modules/solid_properties/src/userobjects/ThermalGraphiteProperties.C]: Then suggested to change the name to `surface_emissivity_from_T`. Naming convention for class member function in MOOSE does not have `_`. Not sure if we should follow it or not here.

**Comment by andrsd** [comment]
> Thoughts? I'll change the design to the `Material`-based design if others agree.

👍 I think this would be much better, because it solves the problem where each property can be a function of arbitrary number of variables much better..

**Comment by YaqiWang** [comment]
I think keeping the interface functions depending only on T is ok. Users can derive the base class in their own way. We derived `TransportFunction` from `Function` before, which has dependency on direction of travel, energy on top of spatial position and time. It looks fine.

**Comment by andrsd** [comment]
If we agree that the only dependency allowed in this module is dependency on T, then I am fine with it. But, I am afraid @aprilnovak was trying to solve a bigger problem.

**Comment by aprilnovak** [comment]
Thanks for the feedback! I will change the design to a `Material`-based structure. A material `ThermalStainlessSteel316` would create properties `cp_solid`, `dcp_solid_dT`, `rho_solid`, `rho_solid_dT`, `k_solid`, and `dk_solid_dT` that could be used directly in applications provided the names match.

It will take some work to update the unit tests, so I won't be able to get to this for a couple days.

**Comment by andrsd** [comment]
> It will take some work to update the unit tests, so I won't be able to get to this for a couple days.

We cannot unit test materials easily right now, so you will need to turn them into normal tests.

**Comment by dschwen** [comment]
> would create properties cp_solid, dcp_solid_dT, rho_solid, rho_solid_dT, k_solid, and dk_solid_dT that could be used directly in applications provided the names match

I would suggest using the `DerivativeMaterialPropertyNameInterface` to generate the derivative names, that way the materials would be valid function materials for use with modules that utilize such derivatives (heat conduction and phase field for example)

**Comment by bwspenc** [comment]
This is a cool idea!  Some thoughts I have about this:
- We need to give some serious thought to what the scope of this should be, and where this kind of capability should live. Not every material model should be in the open-source MOOSE framework for a number of reasons. A set of base classes that arbitrary solid materials can inherit from make a lot of sense in the framework, though.
- Any time you start defining properties for specific materials, unit systems become a big deal. Ideally, something like this would require the user to clearly specify the unit system they want to work with, and base classes would do unit conversions as needed.
- We need to make sure this is interoperable with existing capabilities we have in the heat_conduction and tensor_mechanics modules. We already have a lot of thermal property models like the ones proposed here that define material properties named `density`, `thermal_conductivity`, and `specific_heat`, which is what heat_conduction expects. The names used in this PR are different, so you wouldn't be able to use these models with those modules.

**Comment by aprilnovak** [comment]
> A set of base classes that arbitrary solid materials can inherit from make a lot of sense in the framework, though.

Definitely, I think having a few example materials in the module and appropriate base classes that applications can inherit from would be good. Many of the nuclear applications would probably like to share UO2 properties, but for export control reasons those probably can't go into this module. Maybe there can be another solid properties module hosted privately on the INL gitlab that applications could link to if provided access by INL? 

> base classes would do unit conversions as needed.

I think this would be fairly easy to implement within the base class, we could just provide an enumeration of the different spatial, mass, time, etc. scales. Some properties, such as `ThermalFunctionSolidProperties` could remain dimension-agnostic.

> We need to make sure this is interoperable with existing capabilities we have in the heat_conduction and tensor_mechanics modules.

We could set input parameters in the base classes with user-specified strings for these names, or we could just go through the changes required in all applications by search-and-replace `thermal_conductivity` with whatever name is chosen in this application.

**Comment by aprilnovak** [comment]
This branch has been updated to a materials-based formulation. All materials inherit from `SolidPropertiesMaterial`, which defines every possible method you may want to use across a set of materials.

Specialized classes inherit from `SolidPropertiesMaterial` to compute properties and their derivatives. For example, `ThermalSolidProperties` computes thermal conductivity, density, and specific heat and their derivatives in `computeQpProperties`. Derived classes, such as `ThermalStainlessSteel316Properties`, define the methods to compute those three parameters. A different inheritance path could be `MechanicalSolidProperties`, inheriting from `SolidPropertiesMaterial`, and computing properties such as elastic modulus and Poisson ratio.

I am finding it difficult to incorporate the generality of the user object approach into this materials-based approach. For example: some applications might not need density, specific heat, and thermal conductivity. But, as-written, all applications needing stainless steel density would also need to compute specific heat and thermal conductivity due to inheritance from `ThermalSolidProperties`. 

An alternative design could possibly be as follows. `ThermalStainlessSteel316Properties` inherits from `SolidPropertiesMaterial` instead of `ThermalSolidProperties`. An input parameter to `ThermalSolidProperties` would be a pointer to a `SolidPropertiesMaterial` object, and whatever properties are needed in `ThermalSolidProperties` are computed as `_k[_qp] = _solid_material_ptr->computeThermalConductivity()`. Then, if an application didn't want to compute specific heat, they would simply have a different derived class from `SolidPropertiesMaterial`. So, basically a user-object formulation, but instead of a pointer to a user object, a pointer to a material?

Thoughts?

File: modules/solid_properties/test/tests/functional/test.i [hunk 2]
```diff
@@ -44,22 +44,13 @@
...
-  [../]
-[]
-
 [Materials]
   [./sp_mat]
-    type = ThermalSolidPropertiesMaterial
+    type = ThermalFunctionSolidProperties
     temperature = T
-    sp = sp
+    rho = '1000.0'
```
[the remark this answers is not in the document]
Comment (andrsd) [reply modules/solid_properties/test/tests/functional/test.i]: When we specify a number where function is expected, we create `ConstantFunction` object for it. So, go in `ConstantFunction`, override `timeDerivative` method and let it return 0.

**Comment by tophmatthews** [comment]
Is there a way to start using the ADKernel and ADMaterial capabilities that are new?

**Comment by andrsd** [comment]
This PR is pretty big already. Let's not pile up more stuff here, ok? AD can be added in a separate PR.

**Comment by aprilnovak** [comment]
:(

**Comment by permcody** [comment]
I noticed your frowny face. It's OK - you can always reopen it when you have a chance to revisit. Stale bot can be a bit of a jerk :)

**Comment by andrsd** [comment]
I think she is referring to the fact that it is taking INL forever to finish the paperwork, which we supposedly have to have before we can merge this...

**Comment by aprilnovak** [comment]
yeah, I was referring to the paperwork timeline

**Pull request closed without merging**